KubernetesExecutor multi_namespace_mode can use namespace list to avoid requiring cluster role

XD-DENG opened the pull request:
Currently `KubernetesExecutor`'s `multi_namespace_mode` requires the Scheduler to have cluster-scope role on the Kubernetes Cluster, because it's using function `list_pod_for_all_namespaces()`.

However, in certain enterprise environments, it's not possible for users to have cluster-scope role. For example, they may only get permissions in a namespace, rather on the whole cluster. Always allowing the Scheduler pod to have cluster-scope role is not a good from security aspect either.

This change aims to make `KubernetesExecutor`'s `multi_namespace_mode` work without cluster-scope role.

(This was discussed at the mail list at https://lists.apache.org/thread/xxsppw7qwvky78l6nx41vlz593gj4zqb)

I'm sure folks would have suggestions and we need to future refine this change, but I would like to bring up the discussion by creating this PR first.


**UPDATE:**
Advantages this change brings:
- Better fits enterprise environment
- Better security: limit the permissions that the Scheduler Pod needs, so that it doesn't have too much permissions which it doesn't have to have (earlier it has to have a cluster role in order to use `multi_namespace_mode`)

XD-DENG committed:
Ensure KubernetesExecutor's multi_namespace_mode work without cluster-scope role

Meanwhile also fix a few existing issues.

XD-DENG committed:
Fixup

XD-DENG committed:
Another fixup

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -66,7 +67,7 @@ class ResourceVersion:
 
     _instance = None
-    resource_version = "0"
+    resource_version: dict[str | None, str | None] = {}
```
does this not work?
```suggestion
    resource_version: dict[str, str] = {}
```

XD-DENG replied:
Hi @dstandish , when `multi_namespace_mode` is `True` & `namespace_list` is not specified, that means the user will still want to apply the way using cluster-scope role. In this case, argument `namespace` will be None for the corresponding `KubernetesJobWatcher`. That's they the key can be `None`.

But the value doesn't have to be `str | None`. So I will change this to `resource_version: dict[str | None, str] = {}`.

XD-DENG replied:
Changed to `resource_version: dict[str | None, str] = {}` in commit
https://github.com/apache/airflow/pull/28047/commits/a23c33e90bd6666286e03cc676aba9a6322d570c

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -288,15 +294,33 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
-        else:
-            self.log.error(
-                "Error while health checking kube watcher process. Process died for unknown reasons"
+    def _make_kube_watchers(self) -> dict[str | None, KubernetesJobWatcher]:
```
```suggestion
    def _make_kube_watchers(self) -> dict[str, KubernetesJobWatcher]:
```

XD-DENG replied:
The same as clarified above.

ferruzzi reviewed (approved):
Other than the comments already made, LGTM

dstandish commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -1197,9 +1197,14 @@ delete_worker_pods_on_failure = False
+# If multi_namespace_mode is True while scheduler does not have a cluster-role,
+# give the list of namespaces where the scheduler will schedule jobs
+# Scheduler needs to have the necessary permissions in these namespaces.
+namespace_list =
```
maybe this would be better.  

```suggestion
multi_namespace_mode_namespace_list =
```
it's really a sub-option of multi_namespace_mode.

wdyt?

XD-DENG replied:
Sure, that sounds sensible to me.

Let me add a follow-up commit.

XD-DENG replied:
Address with commit https://github.com/apache/airflow/pull/28047/commits/1c1f25d92e0faeb4f84069056dbdc24c1e53fa93

XD-DENG committed:
Fix static check error

XD-DENG committed:
Address review comment (rename config item)

XD-DENG committed:
Nit fixup for static check

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -140,9 +141,14 @@ def _run(
+            else:
+                list_worker_pods = functools.partial(
+                    watcher.stream, kube_client.list_pod_for_all_namespaces, **kwargs
+                )
```
might be cleaner to move this if block to a private method and i think it can be simplified to this:
```python
    def _list_worker_pods(self, watcher: watch.Watch, kube_client, **kwargs):
        if self.namespace is None:
            return watcher.stream(kube_client.list_pod_for_all_namespaces, **kwargs)
        else:
            return watcher.stream(kube_client.list_namespaced_pod, self.namespace, **kwargs)
```

then used like this
```python
        for event in self._list_worker_pods(watcher, kube_client, **kwargs):
```

I think this makes sense because, this class doesn't really need to know about the airflow config.  It's just a watcher.  And, it's either watching a namespace, or watching all namespaces.  If namespace is none, it watches all; otherwise, it watches the namespace.

XD-DENG replied:
@dstandish  , Great point to skip checking the config and simplify the checks! I have added https://github.com/apache/airflow/pull/28047/commits/2aeb76939d1fdeecc04c1901d85244451de7d2fe to address this.

But IMHO I'm nos sure if we should add one more private method for it, as we don't re-use it anywhere, and passing the args around doesn't seem clean (I did abstract for a similar case in the `KubernetesExecutor` code, the `_list_pods` method, as it's being re-used multiple times)

dstandish replied:
yeah, similar to below, i guess the idea behind the suggestion is that, code reuse isn't the only reason to extract logic into a method. sometimes it just makes the code easier to read... so you just see `for pod in list_pods()` rather than a bunch of logic in an if block.

again, feel free to disregard, just suggestion

side note though, i don't think there's any need to use functools.partial here, i don't think it is doing anything different from just calling stream() here, which returns a generator, and iterating that later.  so i think we can get rid of that (i know that's just how it was)

XD-DENG replied:
Thanks again @dstandish . This part makes sense to me. I added commit https://github.com/apache/airflow/pull/28047/commits/9eaeede8dabc8f823d113ecb85112e43c31a2a8e to introduce the private method `_pod_events`, and get rid of `functool.partial()`. Let me know if you find it ok?

dstandish commented on airflow/kubernetes/kube_config.py:
```diff
@@ -57,6 +57,14 @@ def __init__(self):
         # create, watch, get, and delete pods in this namespace.
         self.kube_namespace = conf.get(self.kubernetes_section, "namespace")
         self.multi_namespace_mode = conf.getboolean(self.kubernetes_section, "multi_namespace_mode")
+        if self.multi_namespace_mode and conf.get(
+            self.kubernetes_section, "multi_namespace_mode_namespace_list"
+        ):
+            self.multi_namespace_mode_namespace_list = conf.get(
+                self.kubernetes_section, "multi_namespace_mode_namespace_list"
+            ).split(",")
+        else:
+            self.multi_namespace_mode_namespace_list = None
```
```suggestion
        self.multi_namespace_mode_namespace_list = conf.getjson(
            self.kubernetes_section, "multi_namespace_mode_namespace_list"
        )
```
maybe this would be a good use of json. or could add a `getlist`

XD-DENG replied:
Hi @dstandish , I'm not sure if this is the correct thing to do. 

I think it's easier to allow users pass a value like `ns-a,ns-b,ns-c` to `multi_namespace_mode_namespace_list`, rather than asking them to put something like `"ns-a","ns-b","ns-c"`. So treating it as a json or list may not be making sense IMHO, and I would suggest we keep the existing way.

Let me know how you find it?

dstandish replied:
sure

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -444,6 +471,23 @@ def __init__(self):
         self.kubernetes_queue: str | None = None
         super().__init__(parallelism=self.kube_config.parallelism)
 
+    def _list_pods(self, query_kwargs):
+        if self.kube_config.multi_namespace_mode:
+            if self.kube_config.multi_namespace_mode_namespace_list:
+                pods = []
+                for namespace in self.kube_config.multi_namespace_mode_namespace_list:
+                    pods.extend(
+                        self.kube_client.list_namespaced_pod(namespace=namespace, **query_kwargs).items
+                    )
+            else:
+                pods = self.kube_client.list_pod_for_all_namespaces(**query_kwargs).items
+        else:
+            pods = self.kube_client.list_namespaced_pod(
+                namespace=self.kube_config.kube_namespace, **query_kwargs
+            ).items
+
+        return pods
+
```
```suggestion
    @cached_property
    def _namespaces(self):
        if self.kube_config.multi_namespace_mode_namespace_list:
            return self.kube_config.multi_namespace_mode_namespace_list
        elif self.kube_config.multi_namespace_mode:
            return []
        else:
            return [self.kube_config.kube_namespace]

    def _list_pods(self, query_kwargs):
        if not self._namespaces:
            yield from self.kube_client.list_pod_for_all_namespaces(**query_kwargs).items
        for namespace in self._namespaces:
            yield from self.kube_client.list_namespaced_pod(namespace=namespace, **query_kwargs).items

```

XD-DENG replied:
Sorry I don't see very obvious improvement/advantage from this change. May you clarify a bit more? Otherwise I believe the existing implementation is clearer IMHO

dstandish replied:
yeah, not really important sometimes i just think generators are a cleaner than managing the data structures

if you want to be really picky you could say that, in the case where you are using a namespace list using the generator will sometimes a few API calls because if, say, you find the pod in the first namespace, you don't need to try the others.

but i think mainly my motivation was just separation of the logic to determine namespaces from the logic to list the pods, because smaller functions can be easier to understand.

but feel free to ignore.

XD-DENG replied:
For this part, I personally find the current coding is easier to comprehend, and no disadvantage in terms performance. So I personally prefer not to make change in this part.

Please let me know if you find it ok? Thanks

dstandish replied:
sure thing

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -497,16 +541,16 @@ def clear_not_launched_queued_tasks(self, session=None) -> None:
             kwargs["label_selector"] += ",run_id=" + pod_generator.make_safe_label_value(ti.run_id)
-            pod_list = self.kube_client.list_namespaced_pod(self.kube_config.kube_namespace, **kwargs)
-            if pod_list.items:
+            pod_list = self._list_pods(kwargs)
+            if pod_list:
```
```suggestion
            for _ in self._list_pods(kwargs):
```
if you accept above suggestion, you'd have to do this

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -497,16 +541,16 @@ def clear_not_launched_queued_tasks(self, session=None) -> None:
             )
-            pod_list = self.kube_client.list_namespaced_pod(self.kube_config.kube_namespace, **kwargs)
-            if pod_list.items:
+            pod_list = self._list_pods(kwargs)
```
and same here

XD-DENG replied:
The lines being changes here seems outdated? They are no longer there in the latest version in this PR.

Regarding the change itself, please refer to my question above.

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -287,15 +293,35 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
+                ResourceVersion().resource_version[namespace] = "0"
+                self.kube_watchers[namespace] = self._make_kube_watcher(namespace)
 
     def run_next(self, next_job: KubernetesJobType) -> None:
```
in this method construct_pod is called with `self.namespace`. does this make sense with multi-namespace mode?  how are additional namespaces used if pods are always created with the one namespace?

XD-DENG replied:
@dstandish  , that's a good question, and thanks for reminding on it.

Actually along preparing this PR, I kept having this question: did this `multi_namespace_mode` feature ever really work for anyone? I also noticed a few similar issues, and I shared them at https://lists.apache.org/thread/nqnq2gbvtx6fyycsqhgjhlz1yotq0vhc

dstandish replied:
It looks like maybe if you don't set `namespace` in airflow cfg then you may be able to set it with executor config

i don't know more than that

are you using it now?

XD-DENG replied:
Our team are using KubernetesExecutor, and we do run it with with a multi namespace setting. But we couldn't use vanilla `multi_namespace_mode` as it is. That's exactly why I'm proposing the changes in this PR ;-)

XD-DENG replied:
> It looks like maybe if you don't set namespace in airflow cfg then you may be able to set it with executor config

And yes, I believe this is the answer to the original question you shared above.

dstandish replied:
cool

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -287,15 +293,35 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
+
+        for namespace in namespaces_to_watch:
+            watchers[namespace] = self._make_kube_watcher(namespace)
+        return watchers
```
this is somewhat redundant logic. maybe you could just calculate "namespaces" once (e.g. in the property i suggested adding to k8s exec) and then pass that along when creating this class.  

then you could do something like 
```
    def _make_kube_watchers(self) -> dict[str | None, KubernetesJobWatcher]:
        if not self.namespaces:
            return {None: self._make_kube_watcher(None)}
        return {x: self._make_kube_watcher(x) for x in self.namespaces}
```

i don't feel great about having `None` as a key in the dict... it's not a super clear as a signaling mechanism. (and,  of course, it's not a namespace, though sometimes it appears that way in the code).

maybe we'll think of something as we work through it.

XD-DENG replied:
Hi @dstandish , regarding the key, I added https://github.com/apache/airflow/pull/28047/commits/928b8399cad70c16c451096cf28a5222c00db084 . The key will be `ALL_NAMESPACES` when `namespace_list` is not given & `multi_namespace_mode=True`. `"ALL_NAMESPACES"` is not legal as K8s namespace, so we don't have to worry about name collision with a real namespace.

Let me know how you find it, comparing with `None`?

Regarding your other comments, will need more time to get back to you (actually haven't got time to read in details yet. Busy with day works). Appreciate the review and the suggestions!

dstandish replied:
> "ALL_NAMESPACES" is not legal as K8s namespace, so we don't have to worry about name collision with a real namespace.

Nice, that's great.  A little more descriptive, and very convenient that there's no risk of collision.  

re the other stuff, no worries...  and... sorry if they are too "code golfy" ... just... was looking at how to maybe streamline some parts of it... just suggestions

XD-DENG replied:
They are definitely not "golfy" I believe! I do appreciate you taking time to help suggest on these details, and I find them extremely useful!

Will get back to you here later.

XD-DENG committed:
use a string as the resource_version key when no namespace_list is provided, rather than using None

We use a string which is illegal for namespace naming, so no risk of name collision

XD-DENG committed:
Simplify the condition checkes when create the watcher stream in JobWatcher._run()

XD-DENG committed:
Address review comment: minor refactor on how to get pod events via stream

dstandish commented on tests/executors/test_kubernetes_executor.py:
```diff
@@ -1108,9 +1209,9 @@ def effect():
                 assert e.args == ("sentinel",)
 
-            # both  resource_version should be 0 after _run raises and exception
+            # both resource_version should be 0 after _run raises and exception
```
```suggestion
            # both resource_version should be 0 after _run raises an exception
```

XD-DENG committed:
Fix a typo tests/executors/test_kubernetes_executor.py

Co-authored-by: Daniel Standish <15932138+dstandish@users.noreply.github.com>

dstandish commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -79,7 +81,7 @@ class KubernetesJobWatcher(multiprocessing.Process, LoggingMixin):
         self,
-        namespace: str | None,
+        namespace: str,
         multi_namespace_mode: bool,
```
i think this can be removed now, along with its attr

XD-DENG replied:
Nice catch! Addressed with https://github.com/apache/airflow/pull/28047/commits/742468ef37a0725d657ac4790f9c3daf22df9766

XD-DENG committed:
JobWatcher doesn't need arg/attr 'multi_namespace_mode' anymore

dstandish reviewed (approved):
looks good to me. maybe worth getting second set of eyes e.g. from @jedcunningham or @uranusjr

uranusjr commented on airflow/executors/kubernetes_executor.py:
```diff
@@ -66,7 +68,7 @@ class ResourceVersion:
     """Singleton for tracking resourceVersion from Kubernetes."""
 
     _instance = None
-    resource_version = "0"
+    resource_version: dict[str, str] = {}
```
We can probably improve this class further, but that can be a separate PR.

uranusjr reviewed (approved):
I didn’t think everything _entirely_ through but this should be good.

XD-DENG committed:
Update test

XD-DENG committed:
Update version_added for multi_namespace_mode_namespace_list

dstandish commented on airflow/config_templates/config.yml:
```diff
@@ -2422,11 +2422,21 @@
+        If multi_namespace_mode is True while scheduler does not have a cluster-role,
+        give the list of namespaces where the scheduler will schedule jobs
+        Scheduler needs to have the necessary permissions in these namespaces.
+      version_added: 2.5.1
```
i think it will need to go in 2.6 since new feature

XD-DENG replied:
Yep I agree. Will update to 2.6.0

(I just found this was missed to be updated so updating it to a "random" value and see if there is any luck it's relevant the weird CI errors. Obviously no luck)

XD-DENG replied:
Updated to 2.6.0 at https://github.com/apache/airflow/pull/28047/commits/974e34521e6d32db7f50ca810c1f602e918f8eec

Continuing to check the CI failures.

XD-DENG committed:
Update version_added for multi_namespace_mode_namespace_list

XD-DENG committed:
End the executor in the test
